altair/utils/core.py

```diff
@@ -41,9 +50,21 @@
     from altair.utils._dfi_types import DataFrame as DfiDataFrame
     from altair.vegalite.v5.schema._typing import StandardType_T as InferredVegaLiteType
 
-V = TypeVar("V")
-P = ParamSpec("P")
 TIntoDataFrame = TypeVar("TIntoDataFrame", bound=IntoDataFrame)
+T = TypeVar("T")
+P = ParamSpec("P")
+R = TypeVar("R")
+
+WrapsFunc = TypeAliasType("WrapsFunc", Callable[..., R], type_params=(R,))
+WrappedFunc = TypeAliasType("WrappedFunc", Callable[P, R], type_params=(P, R))
+# NOTE: Requires stringized form to avoid `< (3, 11)` issues
+# See: https://github.com/vega/altair/actions/runs/10667859416/job/29567290871?pr=3565
```

Comment:
Will this type alias still work as expected in Python < 3.11 or are we just ignoring the error by passing it as a string?

Reply:
So [this error](https://github.com/vega/altair/actions/runs/10667859416/job/29567290871?pr=3565
) (*that is no longer present*) 
<details><summary>Log traceback</summary>
<p>

```py
   File "/home/runner/work/altair/altair/altair/utils/core.py", line 61, in <module>
    "WrapsMethod", Callable[Concatenate[T, ...], R], type_params=(T, R)
  File "/opt/hostedtoolcache/Python/3.10.14/x64/lib/python3.10/typing.py", line 312, in inner
    return func(*args, **kwds)
  File "/opt/hostedtoolcache/Python/3.10.14/x64/lib/python3.10/typing.py", line 403, in __getitem__
    return self._getitem(self, parameters)
  File "/opt/hostedtoolcache/Python/3.10.14/x64/lib/python3.10/typing.py", line 599, in Concatenate
    raise TypeError("The last parameter to Concatenate should be a "
TypeError: The last parameter to Concatenate should be a ParamSpec variable.

```

</p>
</details> 

was being triggered at runtime, but it was not an issue when I used the equivalent as an annotation in an [earlier commit](https://github.com/vega/altair/pull/3565/commits/53057b781ca47349e361a52df50fd322c86a7b32).

Writing it as a string simply gives us the same semantics as writing in an annotation; which seems to be all that was needed in this case.

Reply:
To confirm this works, you can try the [**Repro**](https://github.com/vega/altair/pull/3565#issue-2499636422) on `3.10` which should still be able to use this type to resolve the signatures

Reply:
Got it, thanks :)